scanner/src/funTest/kotlin/experimental/AbstractProvenanceBasedStorageFunTest.kt

```diff
@@ -0,0 +1,239 @@
+            }
+
+            "find scan result for a specific scanner" {
+                val scanResult1 = createScanResult()
+                val scanResult2 = createScanResult(scannerDetails = createScannerDetails(name = "other"))
+
+                storage.write(scanResult1)
+                storage.write(scanResult2)
+
+                val readResult =
+                    storage.read(scanResult1.provenance as KnownProvenance, scanResult1.scanner.toCriteria())
+
+                readResult should containExactlyInAnyOrder(scanResult1)
+            }
+
+            "find all scan results for scanners with names matching a pattern" {
+                val scanResult1 = createScanResult(scannerDetails = createScannerDetails(name = "name1"))
+                val scanResult2 = createScanResult(scannerDetails = createScannerDetails(name = "name2"))
+                val scanResult3 = createScanResult(scannerDetails = createScannerDetails(name = "other name"))
+                val criteria = scanResult1.scanner.toCriteria().copy(regScannerName = "name.+")
+
+                storage.write(scanResult1)
+                storage.write(scanResult2)
+                storage.write(scanResult3)
+
+                val readResult = storage.read(scanResult1.provenance as KnownProvenance, criteria)
+
+                readResult should containExactlyInAnyOrder(scanResult1, scanResult2)
+            }
+
+            "find all scan results for compatible scanners" {
+                val scanResult = createScanResult()
+                val scanResultCompatible1 = createScanResult(scannerDetails = createScannerDetails(version = "1.0.1"))
+                val scanResultCompatible2 =
+                    createScanResult(scannerDetails = createScannerDetails(version = "1.0.1-alpha.1"))
+                val scanResultIncompatible = createScanResult(scannerDetails = createScannerDetails(version = "2.0.0"))
+                val criteria = scanResult.scanner.toCriteria().let { it.copy(maxVersion = it.minVersion.nextMinor()) }
+
+                storage.write(scanResult)
+                storage.write(scanResultCompatible1)
+                storage.write(scanResultCompatible2)
+                storage.write(scanResultIncompatible)
+
+                val readResult = storage.read(scanResult.provenance as KnownProvenance, criteria)
+
+                readResult should containExactlyInAnyOrder(scanResult, scanResultCompatible1, scanResultCompatible2)
+            }
+        }
+    }
+}
+
+private fun ScannerDetails.toCriteria() =
+    ScannerCriteria(
+        regScannerName = name,
+        minVersion = Semver(version),
+        maxVersion = Semver(version).nextPatch(),
+        configMatcher = { true }
+    )
+
+private fun RemoteArtifact.Companion.valid() =
+    RemoteArtifact(
+        url = "https://jitpack.io/com/github/oss-review-toolkit/ort/cli/f42e41a8fe/cli-f42e41a8fe-sources.jar",
```

Comment:
I'm not sure how persistent this URL is, esp if we stop using JitPack at some point when doing real releases on Maven Central. So we should probably use some other artifact here.